src/compiler/resolutionCache.ts

```diff
@@ -0,0 +1,560 @@
+                const { resolvedModule, failedLookupLocations } = loadModuleFromGlobalCache(moduleName, resolutionHost.projectName, compilerOptions, host, globalCache);
+                if (resolvedModule) {
+                    return { resolvedModule, failedLookupLocations: addRange(primaryResult.failedLookupLocations as Array<string>, failedLookupLocations) };
+                }
+            }
+
+            // Default return the result from the first pass
+            return primaryResult;
+        }
+
+        function resolveNamesWithLocalCache<T extends ResolutionWithFailedLookupLocations, R extends ResolutionWithResolvedFileName>(
+            names: string[],
+            containingFile: string,
+            cache: Map<Map<T>>,
+            perDirectoryCache: Map<Map<T>>,
+            loader: (name: string, containingFile: string, options: CompilerOptions, host: ModuleResolutionHost) => T,
+            getResolutionWithResolvedFileName: GetResolutionWithResolvedFileName<T, R>,
+            reusedNames: string[] | undefined,
+            logChanges: boolean): R[] {
+
+            const path = resolutionHost.toPath(containingFile);
+            const resolutionsInFile = cache.get(path) || cache.set(path, createMap()).get(path);
+            const dirPath = getDirectoryPath(path);
+            let perDirectoryResolution = perDirectoryCache.get(dirPath);
+            if (!perDirectoryResolution) {
+                perDirectoryResolution = createMap();
+                perDirectoryCache.set(dirPath, perDirectoryResolution);
+            }
+
+            const resolvedModules: R[] = [];
+            const compilerOptions = resolutionHost.getCompilationSettings();
+
+            const seenNamesInFile = createMap<true>();
+            for (const name of names) {
+                let resolution = resolutionsInFile.get(name);
+                // Resolution is valid if it is present and not invalidated
+                if (!seenNamesInFile.has(name) &&
+                    allFilesHaveInvalidatedResolution || !resolution || resolution.isInvalidated) {
+                    const existingResolution = resolution;
+                    const resolutionInDirectory = perDirectoryResolution.get(name);
+                    if (resolutionInDirectory) {
+                        resolution = resolutionInDirectory;
+                    }
+                    else {
+                        resolution = loader(name, containingFile, compilerOptions, resolutionHost);
+                        perDirectoryResolution.set(name, resolution);
+                    }
+                    resolutionsInFile.set(name, resolution);
+                    const diffIndex = existingResolution && existingResolution.failedLookupLocations && resolution.failedLookupLocations && findDiffIndex(resolution.failedLookupLocations, existingResolution.failedLookupLocations);
+                    watchFailedLookupLocationOfResolution(resolution, diffIndex || 0);

```

Comment:
I would just move the `diffIndex` logic into this function, since `watchFailedLookupLocationOfResolution` *also* checks for `resolution && resolution.failedLookupLocations`.